src/diagram/registry.rs

```diff
@@ -0,0 +1,1421 @@
+        }
+    }
+
+    pub(super) fn scope(&self) -> Entity {
+        self.scope
+    }
+
+    pub(super) fn id(&self) -> Entity {
+        self.target
+    }
+}
+
+impl<T> From<Output<T>> for DynOutput
+where
+    T: Send + Sync + 'static,
+{
+    fn from(output: Output<T>) -> Self {
+        Self {
+            scope: output.scope(),
+            target: output.id(),
+            type_id: TypeId::of::<T>(),
+        }
+    }
+}
+
+/// A type erased [`bevy_impulse::Node`]
+pub(super) struct DynNode {
+    pub(super) input: DynInputSlot,
+    pub(super) output: DynOutput,
+}
+
+impl DynNode {
+    fn new<Request, Response>(output: Output<Response>, input: InputSlot<Request>) -> Self
+    where
+        Request: 'static,
+        Response: Send + Sync + 'static,
+    {
+        Self {
+            input: input.into(),
+            output: output.into(),
+        }
+    }
+}
+
+impl<Request, Response, Streams> From<Node<Request, Response, Streams>> for DynNode
+where
+    Request: 'static,
+    Response: Send + Sync + 'static,
+    Streams: StreamPack,
+{
+    fn from(node: Node<Request, Response, Streams>) -> Self {
+        Self {
+            input: node.input.into(),
+            output: node.output.into(),
+        }
+    }
+}
+
+#[derive(Serialize)]
+pub struct NodeRegistration {
+    pub(super) id: BuilderId,
+    pub(super) name: String,
+    /// type name of the request
+    pub(super) request: &'static str,
+    /// type name of the response
+    pub(super) response: &'static str,
+    pub(super) config_schema: Schema,
+
+    /// Creates an instance of the registered node.
+    #[serde(skip)]
+    create_node_impl: CreateNodeFn,
+}
+
+impl NodeRegistration {
+    pub(super) fn create_node(
+        &self,
+        builder: &mut Builder,
+        config: serde_json::Value,
+    ) -> Result<DynNode, DiagramError> {
+        let n = (self.create_node_impl.borrow_mut())(builder, config)?;
+        debug!(
+            "created node of {}, output: {:?}, input: {:?}",
+            self.id, n.output, n.input
+        );
+        Ok(n)
+    }
+}
+
+type CreateNodeFn =
+    RefCell<Box<dyn FnMut(&mut Builder, serde_json::Value) -> Result<DynNode, DiagramError>>>;
+type DeserializeFn =
+    Box<dyn Fn(&mut Builder, Output<serde_json::Value>) -> Result<DynOutput, DiagramError>>;
+type SerializeFn =
+    Box<dyn Fn(&mut Builder, DynOutput) -> Result<Output<serde_json::Value>, DiagramError>>;
+type ForkCloneFn =
+    Box<dyn Fn(&mut Builder, DynOutput, usize) -> Result<Vec<DynOutput>, DiagramError>>;
+type ForkResultFn =
+    Box<dyn Fn(&mut Builder, DynOutput) -> Result<(DynOutput, DynOutput), DiagramError>>;
+type SplitFn = Box<
+    dyn for<'a> Fn(
+        &mut Builder,
+        DynOutput,
+        &'a SplitOp,
+    ) -> Result<DynSplitOutputs<'a>, DiagramError>,
+>;
+type JoinFn = Box<dyn Fn(&mut Builder, Vec<DynOutput>) -> Result<DynOutput, DiagramError>>;
+
+pub struct CommonOperations<'a, Deserialize, Serialize, Cloneable> {
```

Comment:
Unrelated to this PR but we should probably add `#[must_use]` to this struct.

Reply:
6cf280988df4a468bce851037a41e430e9227ab7..4e6d8ac6f8c2fed0de962a38ab56016f72634eb2